src/pcnt.rs

```diff
@@ -0,0 +1,543 @@
+        match self {
+            PcntChannel::Channel0 => esp_idf_sys::pcnt_channel_t_PCNT_CHANNEL_0,
+            PcntChannel::Channel1 => esp_idf_sys::pcnt_channel_t_PCNT_CHANNEL_1,
+        }
+    }
+}
+
+#[doc = " @brief PCNT channel action on signal edge"]
+#[derive(Debug, Copy, Clone)]
+pub enum PcntCountMode {
+    #[doc = "< Hold current count value"]
+    Hold,
+    #[doc = "< Increase count value"]
+    Increment,
+    #[doc = "< Decrease count value"]
+    Decrement,
+}
+
+impl Into<esp_idf_sys::pcnt_count_mode_t> for PcntCountMode {
+    fn into(self) -> esp_idf_sys::pcnt_count_mode_t {
+        match self {
+            PcntCountMode::Hold => esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_HOLD,
+            PcntCountMode::Increment => esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_INCREASE,
+            PcntCountMode::Decrement => esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_DECREASE,
+        }
+    }
+}
+
+#[doc = " @brief PCNT channel action on control level"]
+#[derive(Debug, Copy, Clone)]
+pub enum PcntControlMode {
+    #[doc = "< Keep current count mode"]
+    Keep,
+    #[doc = "< Invert current count mode (increase -> decrease, decrease -> increase)"]
+    Reverse,
+    #[doc = "< Hold current count value"]
+    Disable,
+}
+
+impl Into<esp_idf_sys::pcnt_ctrl_mode_t> for PcntControlMode {
+    fn into(self) -> esp_idf_sys::pcnt_ctrl_mode_t {
+        match self {
+            PcntControlMode::Keep => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_KEEP,
+            PcntControlMode::Reverse => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_INVERSE,
+            PcntControlMode::Disable => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_HOLD,
+        }
+    }
+}
+
+bitflags! {
+    #[allow(dead_code)]
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct PcntEventType: u32 {
+        #[doc = "< PCNT watch point event: threshold1 value event"]
+        const THRES_1 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_1;
+        #[doc = "< PCNT watch point event: threshold0 value event"]
+        const THRES_0 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_0;
+        #[doc = "< PCNT watch point event: Minimum counter value"]
+        const L_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_L_LIM;
+        #[doc = "< PCNT watch point event: Maximum counter value"]
+        const H_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_H_LIM;
+        #[doc = "< PCNT watch point event: counter value zero event"]
+        const ZERO = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_ZERO;
+    }
+}
+
+#[doc = " @brief Pulse Counter configuration for a single channel"]
+pub struct PcntConfig<'d> {
+    #[doc = "< Pulse input GPIO number, if you want to use GPIO16, enter pulse_gpio_num = 16, a negative value will be ignored"]
+    pub pulse_pin: Option<&'d AnyInputPin>,
```

Comment:
* You are not supposed to require the user to downgrade a pin, and you are doing this here, by using `AnyOutputPin`. You should be using a generic instead
* You are not supposed to impose to the user whether the pin is passed by reference, or by value. This is done with the `impl Peripheral<P = impl InputPin>` magic
* Since the above magic is difficult to use in a struct (results in a heavily generified struct), the pin peripheral should not be part of the config. We are not doing this anyway - peripherals are always passed directly to the `new` constructor of the driver, and never part of the config.

Reply:
updated to use impl Peripheral<P = impl InputPin> magic